### PR TITLE
stm32h7: Transition clock configuration from Kconfig to device tree

### PR DESCRIPTION
**Pull request by erwango**
Similarly to https://github.com/zephyrproject-rtos/zephyr/pull/34120, convert stm32h7 clock_control driver to the use of device tree for clocks configuration

EDIT:
Includes https://github.com/zephyrproject-rtos/zephyr/pull/34704 to avoid CI related issues

### CODE REVIEW

**Comment by FRASTM** [comment]
About H7 and clock device tree there any link with the #34691, I mean is it one still valid after "_stm32h7: Transition clock configuration from Kconfig to device tree_"

**Comment by erwango** [comment]
> About H7 and clock device tree there any link with the #34691, I mean is it one still valid after "_stm32h7: Transition clock configuration from Kconfig to device tree_" ?

The one that gets merged first will need to take the other one into account.
For now, https://github.com/zephyrproject-rtos/zephyr/pull/34691 is missing some important bits like code using these symbols.

File: boards/arm/nucleo_h745zi_q/nucleo_h745zi_q.dtsi [hunk 1]
```diff
@@ -28,3 +28,12 @@
 		};
 	};
 };
+
+&rcc {
```
Comment (FRASTM) [inline boards/arm/nucleo_h745zi_q/nucleo_h745zi_q.dtsi:32]: we could retrieve the comment like the in the commit boards/arm/nucleo_h723zg/nucleo_h723zg.dts   does, for example
`hpre = <2>;    /* HCLK: 240   MHz */`

File: dts/arm/st/h7/stm32h7.dtsi [hunk 2]
```diff
@@ -38,6 +39,61 @@
...
+			/* H7 clock driver may not always need this value */
+			/* but it is required by the binding */
+			clock-frequency = <DT_FREQ_M(25)>;
+			status = "disabled";
+		};
+
+		clk_hsi: clk-hsi {
+			#clock-cells = <0>;
+			compatible = "st,stm32h7-hsi-clock";
+			status = "disabled";
```
Comment (FRASTM) [inline dts/arm/st/h7/stm32h7.dtsi:58]: maybe `		clock-frequency = <DT_FREQ_M(64)>;`

File: dts/arm/st/h7/stm32h7.dtsi [hunk 3]
```diff
@@ -38,6 +39,61 @@
...
+	clocks {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		clk_hse: clk-hse {
+			#clock-cells = <0>;
+			compatible = "st,stm32-hse-clock";
+			/* H7 clock driver may not always need this value */
+			/* but it is required by the binding */
+			clock-frequency = <DT_FREQ_M(25)>;
```
Comment (FRASTM) [inline dts/arm/st/h7/stm32h7.dtsi:51]: why setting 25MHz then ?  Even if not used here, most of the stm32h7  have `clock-frequency = <DT_FREQ_M(8)>;``

File: dts/bindings/clock/st,stm32h7-rcc.yaml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      required: false
+      type: int
+      description: |
+        default frequency in Hz for clock output
+
+    d1cpre:
+      type: int
+      required: false
+      enum:
+        - 1
```
Comment (ABOSTM) [inline dts/bindings/clock/st,stm32h7-rcc.yaml:63]: d1cpre:  /1 /2 /4 /8 /16 /63 /128 /256

File: dts/bindings/clock/st,stm32h7-rcc.yaml [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      required: false
+      description: |
+          D2 domain, CPU2 core clock and AHB(1/2/3/4) peripheral prescaler
+      enum:
+        - 1
+        - 2
+        - 4
+        - 8
+        - 16
+        - 32
```
Comment (ABOSTM) [inline dts/bindings/clock/st,stm32h7-rcc.yaml:82]: /32 is not possible

File: include/drivers/clock_control/stm32_clock_control.h [hunk 6]
```diff
@@ -75,18 +75,37 @@
...
 	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32l4_pll_clock, okay) || \
-	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32wb_pll_clock, okay)
+	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32wb_pll_clock, okay) || \
+	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32h7_pll_clock, okay)
 #define STM32_PLL_M_DIVISOR	DT_PROP(DT_NODELABEL(pll), div_m)
 #define STM32_PLL_N_MULTIPLIER	DT_PROP(DT_NODELABEL(pll), mul_n)
 #define STM32_PLL_P_DIVISOR	DT_PROP(DT_NODELABEL(pll), div_p)
 #define STM32_PLL_Q_DIVISOR	DT_PROP(DT_NODELABEL(pll), div_q)
 #define STM32_PLL_R_DIVISOR	DT_PROP(DT_NODELABEL(pll), div_r)
+#warning "ho ho"
```
Comment (ABOSTM) [inline include/drivers/clock_control/stm32_clock_control.h:108]: Remove warning

File: dts/arm/st/h7/stm32h7.dtsi (hunk 2 above)
Comment (erwango) [reply dts/arm/st/h7/stm32h7.dtsi:58]: Right

File: boards/arm/nucleo_h745zi_q/nucleo_h745zi_q.dtsi (hunk 1 above)
Comment (erwango) [reply boards/arm/nucleo_h745zi_q/nucleo_h745zi_q.dtsi:32]: Sure

File: dts/arm/st/h7/stm32h7.dtsi (hunk 3 above)
Comment (erwango) [reply dts/arm/st/h7/stm32h7.dtsi:51]: 25MHz is value of HSE_VALUE provided by Cube HAL

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: include/drivers/clock_control/stm32_clock_control.h (hunk 6 above)
Comment (erwango) [reply include/drivers/clock_control/stm32_clock_control.h:108]: Don't like it :-)?

File: dts/bindings/clock/st,stm32h7-rcc.yaml (hunk 4 above)
Comment (erwango) [reply dts/bindings/clock/st,stm32h7-rcc.yaml:63]: Only 1 is possible today:
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/clock_control/clock_stm32_ll_h7.c#L149

File: dts/bindings/clock/st,stm32h7-rcc.yaml (hunk 5 above)
Comment (erwango) [reply dts/bindings/clock/st,stm32h7-rcc.yaml:82]: This one was tricky, txs

File: include/drivers/clock_control/stm32_clock_control.h (hunk 6 above)
Comment (ABOSTM) [reply include/drivers/clock_control/stm32_clock_control.h]: Christmas is now far away :-) .
I like comment if message is more explicit.

File: dts/bindings/clock/st,stm32h7-rcc.yaml (hunk 4 above)
Comment (ABOSTM) [reply dts/bindings/clock/st,stm32h7-rcc.yaml:63]: OK, that is strange to have something configurable, ... but with a single value allowed :-(
Ready for further improvement.

**Review by ABOSTM — APPROVED** [review]
LGTM

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Commit message typo :
board: nucleo_**h723**zg: Convert to dts based clock configuration

**Comment by erwango** [comment]
> Commit message typo :
> board: nucleo_**h723**zg: Convert to dts based clock configuration

Fixed, txs

**Review by ABOSTM — APPROVED** [review]
LGTM

**Review by mbolivar-nordic — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**